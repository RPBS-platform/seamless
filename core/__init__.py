from .cell import cell, pythoncell
from .manager import connect
from .macro import macro
