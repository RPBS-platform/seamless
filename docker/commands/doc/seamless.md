% seamless(1) Version 0.1 | seamless-cli Documentation

NAME
====

**seamless** - Starts a new Seamless Docker container and runs a command inside it.

SYNOPSIS
========

| **seamless** \[_command_]

DESCRIPTION
===========

Starts a new Seamless Docker container and runs a command inside it.

The current directory is mounted to /cwd, and the command is executed there

/tmp is mounted as well

Uses the host network for the Docker container. Will only work under Linux.

BUGS
====

See GitHub Issues: <https://github.com/sjdv1982/seamless/issues>

AUTHOR
======

Sjoerd de Vries <sjdv1982@gmail.com>

SEE ALSO
========

**seamless-bash(1)**, **seamless-ipython(1)**, **seamless-jupyter(1)**