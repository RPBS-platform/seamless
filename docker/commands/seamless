set -u -e

firstarg=$1 #there must be at least one argument
gdbflags='--cap-add=SYS_PTRACE --security-opt seccomp=unconfined --security-opt apparmor=unconfined'
docker run --rm \
  $gdbflags \
  --network=host \
  -v `pwd`:/cwd \
  -v /tmp:/tmp \
  --workdir /cwd \
  --user $(id -u):$(id -g) \
  rpbs/seamless $*
