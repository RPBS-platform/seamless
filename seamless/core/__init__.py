import weakref
from enum import Enum

class IpyString(str):
    def _repr_pretty_(self, p, cycle):
        return p.text(str(self))

class SeamlessBase:
    _destroyed = False
    _macro_object = None # macro object that CREATED this instance
    _context = None
    _last_context = None
    name = None

    StatusFlags = Enum('StatusFlags', ('OK', 'PENDING', 'UNDEFINED', 'UNCONNECTED', 'ERROR'))
    _status = StatusFlags.UNDEFINED

    @property
    def path(self):
        if self._context is None:
            return ()
        elif self._context.path is None:
            return ("<None>", self.name)
        else:
            return self._context.path + (self.name,)

    def _validate_path(self, required_path=None):
        if required_path is None:
            required_path = self.path
        else:
            assert self.path == required_path, (self.path, required_path)
        return required_path


    def _set_context(self, context, name, force_detach=False):
        from .context import Context
        assert isinstance(context, Context)
        if self._context is not None:
            assert self.name is not None
            if context is not self._context or force_detach:
                childname = self.name
                assert self._context._children[childname] is self
                self._context._children.pop(childname)
                if childname in self._context._auto:
                    self._context._auto.remove(childname)
        if context is not None:
            self._last_context = context
        self._context = context
        self.name = name
        return self

    @property
    def context(self):
        return self._context


    def _macro_control(self):
        if self._macro_object is not None:
            return self

        ret = None
        if self.context is not None:
            ret = self.context._macro_control()
        else:
            return None

    def format_path(self):
        if self.path is None:
            ret = "<None>"
        else:
            ret = "." + ".".join(self.path)
        return ret

    def __str__(self):
        ret = "Seamless object: " + self.format_path()
        return ret

    def __repr__(self):
        return self.__str__()

    @property
    def macro(self):
        """Returns the macro object associated with this construct"""
        return self._macro_object
    def _set_macro_object(self, macro_object):
        self._macro_object = macro_object

#from .cell import Cell ###
#from .worker import Worker ###
from .context import Context, context ###
