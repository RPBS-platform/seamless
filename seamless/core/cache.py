import functools
import weakref

use_caching = True

@functools.lru_cache(10000)
def _long_signature_cell(cell, root):
    if isinstance(cell, (Inchannel, Outchannel)):
        return None #uncachable for now
    result = {}
    mgr = cell._get_manager()
    from_pin = mgr.cell_from_pin.get(cell)
    if from_pin is not None:
        from_pin = from_pin[1]
    from_cell = mgr.cell_from_cell.get(cell)
    if from_cell is not None:
        from_cell = from_cell[1]
    if from_pin is not None:
        worker = from_pin.worker_ref()
        if worker._context()._part_of(root):
            assert isinstance(worker, Transformer) #TODO: reactors
            #TODO: for reactors, also indicate which outputpin; discount editpins
            result["origin"] = "transformer"
            result["dependency"] = _long_signature_transformer(worker, root)
        else:
            result["origin"] = "extern"
            result["path"] = from_pin.path
    elif from_cell is not None:
        if from_cell._context()._part_of(root):
            result["origin"] = "cell"
            result["dependency"] = _long_signature_cell(from_cell, root)
        else:
            result["origin"] = "extern"
            result["path"] = from_cell.path
    else: #source cell
        if cell.status() == "OK":
            result["origin"] = "self"
            result["checksum"] = cell.checksum()
            #TODO: sometimes it is the text checksum we are interested in
            # Depends on the nature of the downstream connection
            # So we would have separate versions of the long_signature of this cell
        else:
            result["origin"] = "none"
    return result

@functools.lru_cache(10000)
def _long_signature_transformer(tf, root):
    result = {}
    mgr = tf._get_manager()
    for pinname, pin in tf._pins.items():
        if not isinstance(pin, InputPinBase):
            continue
        from_cell = mgr.pin_from_cell.get(pin)
        if from_cell is not None:
            from_cell = from_cell[1]
        subresult = {}
        if from_cell is not None:
            ctx = from_cell._context()
            if ctx._part_of(root):
                subresult["origin"] = "cell"
                subresult["dependency"] = _long_signature_cell(from_cell, root)
            else:
                subresult["origin"] = "extern"
                subresult["path"] = from_cell.path
        else:
            subresult["origin"] = "none"
        result[pinname] = subresult
    return result

def long_signature(obj, root):
    """Calculates the long signature of a cell or worker
    It takes into account the direct and indirect upstream dependencies
    Dependencies above root are considered as extern (and their path is stored)"""
    if isinstance(obj, Cell):
        return _long_signature_cell(obj, root)
    elif isinstance(obj, Transformer):
        return _long_signature_transformer(obj, root)
    else:
        raise TypeError()

def _short_signature_cell(cell):
    if isinstance(cell, (Inchannel, Outchannel)):
        return None #TODO
    if cell.status() == "OK":
        return cell.checksum()
        #TODO: sometimes it is the text checksum we are interested in
        # Depends on the nature of the downstream connection
        # So we would have separate versions of the short_signature of this cell
    else:
        return None

def _short_signature_transformer(tf):
    if tf.status() != "OK":
        return None
    result = {}
    mgr = tf._get_manager()
    for pinname, pin in tf._pins.items():
        if not isinstance(pin, InputPinBase):
            continue
        from_cell = mgr.pin_from_cell.get(pin)
        if from_cell is not None:
            from_cell = from_cell[1]
        subresult = _short_signature_cell(from_cell)
        ###assert subresult is not None #if status is OK, no connected cell can be None; WRONG in case of channels!
        result[pinname] = subresult
    return result

def short_signature(obj):
    """Calculates the short signature of a cell or worker
    It takes into account the direct upstream dependencies
    Source or non-source is not a factor, nor is extern/non-extern
    If any of them is None, the short signature is None as well"""
    if isinstance(obj, Cell):
        return _short_signature_cell(obj)
    elif isinstance(obj, Transformer):
        return _short_signature_transformer(obj)
    else:
        raise TypeError()

def _do_cache(new_obj, old_obj, hits):
    if isinstance(new_obj, Cell):
        assert isinstance(old_obj, Cell)
        new_obj._val = old_obj._val
        new_obj._last_checksum = old_obj._last_checksum
        new_obj._last_text_checksum = old_obj._last_text_checksum
        new_obj._exception = old_obj._exception
        new_obj._prelim_val = old_obj._prelim_val
        new_obj._authoritative = old_obj._authoritative
        new_obj._overruled = old_obj._overruled
        old_obj._val = None
        hits["cells"][new_obj] = old_obj
    elif isinstance(new_obj, Transformer):
        assert isinstance(old_obj, Transformer)
        t = old_obj.transformer
        #TODO: thread-safe lock mechanism to modify t atomically
        t.parent = weakref.ref(new_obj)
        t.send_message("@RESTART", None)
        t.output_queue = new_obj.output_queue
        t.output_semaphore = new_obj.output_semaphore
        old_obj.output_thread.join() #happens quickly after @RESTART signal
        new_obj._listen_output_state = old_obj._listen_output_state
        new_obj.transformer = t
        new_obj.transformer_thread = old_obj.transformer_thread
        new_obj._pending_updates = old_obj._pending_updates
        new_obj._last_value = old_obj._last_value
        new_obj._last_value_preliminary = old_obj._last_value_preliminary
        new_obj._message_id = old_obj._message_id
        assert not new_obj.active
        new_obj._last_update_checksums = old_obj._last_update_checksums.copy()
        old_obj.active = False
        old_obj.transformer = None
        old_obj.transformer_thread = None
        old_obj.destroy()
        hits["transformers"][new_obj] = old_obj
    else:
        raise TypeError(new_obj)


def _cache(obj, root, old_paths, hits):
    path = obj.path
    long_sig = long_signature(obj, root)
    if path in old_paths and old_paths[path][0] == long_sig:
        #print("cache hit by preservation", obj)
        p = old_paths.pop(path)
        return _do_cache(obj, p[2], hits)
    for old_path in list(old_paths.keys()):
        p = old_paths[old_path]
        if p[0] == long_sig:
            #print("cache hit by long signature", obj, old_path)
            old_paths.pop(old_path)
            return _do_cache(obj, p[2], hits)
    short_sig = short_signature(obj)
    if short_sig is not None:
        for old_path in list(old_paths.keys()):
            p = old_paths[old_path]
            if p[1] == short_sig:
                #print("cache hit by short signature", obj, old_path)
                old_paths.pop(old_path)
                return _do_cache(obj, p[2], hits)
    #print("cache miss", obj)

def cache(ctx, old_ctx):
    hits = {"transformers": {}, "cells": {}}
    if not use_caching:
        return hits
    old_paths = {}
    def build_paths(c):
        for child in c._children.values():
            if isinstance(child, Context):
                build_paths(child)
            elif isinstance(child, (Cell, Transformer)): #TODO: reactors
                if isinstance(child, (Inchannel, Outchannel)):
                    continue
                path = child.path
                long_sig = long_signature(child, old_ctx)
                short_sig = short_signature(child)
                old_paths[path] = long_sig, short_sig, child
    build_paths(old_ctx)
    def walk(c):
        for child in c._children.values():
            if isinstance(child, Context):
                walk(child)
            elif isinstance(child, (Cell, Transformer)): #TODO: reactors
                _cache(child, ctx, old_paths, hits)
    walk(ctx)
    return hits

from . import Context, Cell, Transformer #TODO: reactors
from .worker import InputPinBase
from .structured_cell import Inchannel, Outchannel

"""
Something to consider (long term): topology hits
  Macro caching can give a topology hit (if the macro code and macro args are the same).
  This is different from the current approach of value hits, which depend on the value of the connected cells)
  Topology hits may interfere with layers, but maybe they could be made to work
"""
