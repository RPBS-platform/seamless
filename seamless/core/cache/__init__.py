class CacheMissError(Exception):
    pass

from .cell_cache import CellCache
from .accessor_cache import AccessorCache, Accessor
from .expression_cache import ExpressionCache, Expression
from .value_cache import ValueCache, SemanticKey
from .label_cache import LabelCache
from .transform_cache import TransformCache
from .tempref import TempRefManager