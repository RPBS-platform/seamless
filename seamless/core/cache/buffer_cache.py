import time
import weakref
import traceback
from weakref import WeakValueDictionary
import functools
from collections import namedtuple

from .database_client import database_sink, database_cache

# Note: non-authoritative buffers are not cached locally, but they are sent into the database sink
LIFETIME_TEMP = 20.0 # buffer_cache keeps buffer values alive for 20 secs
LIFETIME_TEMP_SMALL = 3600.0 # buffer_cache keeps small buffer values (< 10 000 bytes) alive for an hour
LIFETIME_TF_RESULT = 20.0 # buffer_transformation_result keeps remote/dummy transformation results alive for 20 secs
LIFETIME_TF_RESULT_SMALL = 3600.0 # buffer_transformation_result keeps small (< 10 000 bytes) remote/dummy transformation results alive for an hour

import logging
logger = logging.getLogger("seamless")

def print_info(*args):
    msg = " ".join([str(arg) for arg in args])
    logger.info(msg)

def print_warning(*args):
    msg = " ".join([str(arg) for arg in args])
    logger.warning(msg)

def print_debug(*args):
    msg = " ".join([str(arg) for arg in args])
    logger.debug(msg)

def print_error(*args):
    msg = " ".join([str(arg) for arg in args])
    logger.error(msg)

class BufferCache:
    """Checksum-to-buffer cache.
    Every buffer is referred to by a CacheManager (or more than one).

    Memory intensive. Like any other cache, does not persist unless offloaded.
    Cache items are directly serializable, and can be shared over the
    network, or offloaded to a database.
    Keys are straightforward buffer checksums.

    NOTE: if there is a database sink, buffers are not maintained in local cache.
    Refcounts are still maintained.
    """
    def __init__(self):
        self.buffer_cache = {} #local cache, checksum-to-buffer
        self.last_time = {}
        self.buffer_refcount = {} #buffer-checksum-to-refcount
        # Buffer length cache (never expire)
        self.buffer_length = {} #checksum-to-bufferlength
        self.non_authoritative = set()
        self.missing = set()

    def _check_delete_buffer(self, checksum):
        if checksum not in self.last_time:
            return
        t = time.time()
        l = self.buffer_length.get(checksum, 999999999)
        lifetime = LIFETIME_TEMP_SMALL if l < 10000 else LIFETIME_TEMP
        last_time = self.last_time[checksum]
        curr_lifetime = t - last_time
        if curr_lifetime < lifetime:
            func = functools.partial(self._check_delete_buffer, checksum)
            delay = max(lifetime-curr_lifetime, 1)
            temprefmanager.add_ref(func, delay, on_shutdown=False)
            return

        self.last_time.pop(checksum)
        if checksum in self.buffer_refcount:
            local = (not database_sink.active) or (not database_cache.active)
            if local and checksum not in self.non_authoritative:
                return
        self.buffer_cache.pop(checksum, None)

    def _update_time(self, checksum, buffer_length=None):
        t = time.time()
        if buffer_length is None:
            buffer_length = 9999999
        if checksum not in self.last_time:
            func = functools.partial(self._check_delete_buffer, checksum)
            delay = LIFETIME_TEMP_SMALL if buffer_length < 10000 else LIFETIME_TEMP
            temprefmanager.add_ref(func, delay, on_shutdown=False)
        self.last_time[checksum] = t

    def cache_buffer(self, checksum, buffer):
        """Caches a buffer locally for a short time, without incrementing its refcount
        Does not write into the database.
        If the buffer already has a refcount AND a database is active, nothing happens
        """
        assert checksum is not None
        assert isinstance(buffer, bytes)
        #print("LOCAL CACHE", checksum.hex())
        local = (not database_sink.active) or (not database_cache.active)
        if not local and checksum in self.buffer_refcount:
            return
        self._update_time(checksum, len(buffer))
        if checksum not in self.buffer_cache:
            self.buffer_cache[checksum] = buffer

        if checksum in self.missing:
            self.missing.discard(checksum)
            if database_sink.active:
                database_sink.set_buffer(checksum, buffer, authoritative)

    def incref_buffer(self, checksum, buffer, authoritative):
        """Caches a buffer and increments its refcount.
        If it is the first reference, it is written into the database;
          if there is no database, it is written into local cache instead
        If it is not the first reference but the first authoritative one,
         the buffer is re-written into the database"""
        assert checksum is not None
        assert isinstance(buffer, bytes)
        l = len(buffer)
        if checksum not in self.buffer_length:
            self.buffer_length[checksum] = l
            database_sink.set_buffer_length(checksum, l)
        self._incref(checksum, authoritative, buffer)

    def _incref(self, checksum, authoritative, buffer):
        #print("INCREF     ", checksum.hex(), authoritative, buffer is None)
        if checksum in self.buffer_refcount:
            self.buffer_refcount[checksum] += 1
            if database_sink.active:
                if authoritative and checksum in self.non_authoritative:
                    self.non_authoritative.discard(checksum)
                    if buffer is None:
                        buffer = self.get_buffer(checksum)
                    if buffer is not None:
                        # TODO: this will normally not work. Add a database_sink "make_authoritative" API function!
                        database_sink.set_buffer(checksum, buffer, authoritative)
            if buffer is not None and checksum in self.missing:
                self.missing.discard(checksum)
                local = (not database_sink.active) or (not database_cache.active)
                if authoritative and local:
                    if checksum not in self.buffer_cache:
                        self.buffer_cache[checksum] = buffer
        else:
            self.buffer_refcount[checksum] = 1
            local = (not database_sink.active) or (not database_cache.active)
            if not authoritative:
                self.non_authoritative.add(checksum)
            if buffer is None:
                buffer = self.buffer_cache.get(checksum)
            if buffer is not None:
                if database_sink.active:
                    database_sink.set_buffer(checksum, buffer, authoritative)
                if local:
                    if authoritative:
                        if checksum not in self.buffer_cache:
                            self.buffer_cache[checksum] = buffer
                    else:
                        self.cache_buffer(checksum, buffer)
            else:
                print("MISS", checksum.hex(), checksum in self.buffer_cache)
                raise Exception ###
                self.missing.add(checksum)
            if not local and checksum in self.last_time:
                self.last_time.pop(checksum)
                self.buffer_cache.pop(checksum)

    def incref(self, checksum, authoritative):
        """Increments the refcount of a buffer checksum.

        If it is the first authoritative reference, it is re-written into the database
        (if the buffer can be retrieved)"""

        buffer = None
        if checksum not in self.buffer_refcount:
            buffer = self.buffer_cache.get(checksum)
        return self._incref(checksum, authoritative, buffer)

    def decref(self, checksum):
        """Decrements the refcount of a buffer checksum, cached with incref_buffer
        If the refcount reaches zero, and there is no database,
         it will be added to local cache using cache_buffer.
        This means that it will remain accessible for a short while
        """
        #print("DECREF     ", checksum.hex())
        if checksum not in self.buffer_refcount:
            print_warning("double decref, %s" % checksum.hex())
            return
        self.buffer_refcount[checksum] -= 1
        if self.buffer_refcount[checksum] == 0:
            self.buffer_refcount.pop(checksum)
            self.missing.discard(checksum)
            local = (not database_sink.active) or (not database_cache.active)
            #print("DESTROY", checksum.hex(), local, checksum in self.buffer_cache)
            if local:
                buffer = self.get_buffer(checksum)
                if buffer is not None:  # should be ok normally
                    self.cache_buffer(checksum, buffer)

    def get_buffer(self, checksum):
        if checksum is None:
            return None
        buffer = self.buffer_cache.get(checksum)
        if buffer is not None:
            return buffer
        return database_cache.get_buffer(checksum)

    def get_buffer_length(self, checksum):
        if checksum is None:
            return None
        length = self.buffer_length.get(checksum)
        if length is not None:
            return length
        length = database_cache.get_buffer_length(checksum)
        if length is not None:
            return length
        buf = self.get_buffer(checksum)
        if buf is not None:
            return len(buf)

    def buffer_check(self, checksum):
        """For the communion_server..."""
        assert checksum is not None
        if checksum in self.buffer_cache:
            return True
        return database_cache.has_buffer(checksum)

    def destroy(self):
        if self.buffer_cache is None:
            return
        if len(self.buffer_refcount):
            print_warning("buffer cache, %s buffers undestroyed" % len(self.buffer_refcount))
        self.buffer_cache = None
        self.last_time = None
        self.buffer_refcount = None
        self.buffer_length = None
        self.non_authoritative = None


buffer_cache = BufferCache()
buffer_cache.database_cache = database_cache
buffer_cache.database_sink = database_sink

from ..protocol.calculate_checksum import checksum_cache
from .tempref import temprefmanager
from . import CacheMissError