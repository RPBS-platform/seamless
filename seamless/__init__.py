"""
Seamless: framework for data-driven and live programming
Copyright 2016-2018, Sjoerd de Vries
"""

import sys
import time
import functools
import traceback
import atexit

from abc import abstractmethod
class Wrapper:
    @abstractmethod
    def _unwrap(self):
        pass

from . import silk
from . import mixed

#Dependencies of seamless

# 1. hard dependencies; without these, "import seamless" will fail.
# Still, if necessary, some of these dependencies could be removed, but seamless would have to be more minimalist in loading its lib

import numpy as np

if np.dtype(np.object).itemsize != 8:
    raise ImportError("Seamless requires a 64-bit system")

#silk must be imported before mixed
from . import silk
from . import mixed

from .core import mainloop as _mainloop
from .core.mainloop import mainloop, mainloop_one_iteration, asyncio_finish, workqueue
atexit.register(asyncio_finish)

ipython = None
ipy_error = None
try:
    from IPython import get_ipython
    from IPython.core.error import UsageError
    from IPython.terminal.pt_inputhooks import register as _register_integration_terminal
    from ipykernel.eventloops import register_integration as _register_integration_kernel
except ImportError:
    raise
    ipy_error = "Cannot find IPython"
else:
    ipython = get_ipython()
    if ipython is None:
        ipy_error = "Seamless was not imported inside IPython"

if ipy_error is None:
    last_exception = None
    def new_except_hook(etype, evalue, tb):
        global last_exception
        exc = traceback.format_exception(etype, evalue, tb)
        if exc != last_exception:
            last_exception = exc
            print("".join(exc))

    def patch_excepthook():
        sys.excepthook = new_except_hook
    patch_excepthook()

    def mainloop():
        raise RuntimeError("Cannot run seamless.mainloop() in IPython mode")

    def inputhook_terminal(context):
        while not context.input_is_ready():
            mainloop_one_iteration()
    _register_integration_terminal("seamless", inputhook_terminal)

    @_register_integration_kernel('seamless')
    def inputhook_kernel(kernel):
        while 1:
            t = time.time()
            while time.time() - t < kernel._poll_interval:
                mainloop_one_iteration()        
            kernel.do_one_iteration()

    _register_integration_terminal("seamless", inputhook_terminal)
    
    ipython.enable_gui("seamless")
    
else:
    sys.stderr.write("    " + ipy_error + "\n")
    sys.stderr.write("    Call seamless.mainloop(), seamless.flush() or context.equilibrate() to process cell updates\n")

def flush():
    from .core.mainloop import workqueue
    workqueue.flush()

from .highlevel import *
from .silk import Silk
from .debugger import pdb
