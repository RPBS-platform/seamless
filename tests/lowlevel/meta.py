from seamless.core import context, transformer
from seamless.core.cache.buffer_cache import buffer_cache
import json
ctx = context(toplevel=True)
ctx.tf = transformer({
    "a": {
        "io": "input",
        "celltype": "int",
    },
    "result": {
        "io": "output",
        "celltype": "int",
    },

})
ctx.tf.code.cell().set("a + 42")
ctx.tf.a.cell().set(1)
ctx.tf.meta = {
    "calculation_time": "1s"
}
ctx.compute()

# Adapted from tasks/transformer_update.py
manager = ctx._get_manager()
livegraph = manager.livegraph
taskmanager = manager.taskmanager
upstreams = livegraph.transformer_to_upstream[ctx.tf]

inputpin_checksums = {}
for pinname, accessor in upstreams.items():
    inputpin_checksums[pinname] = accessor._checksum

celltypes = {}
for pinname, accessor in upstreams.items():
    wa = accessor.write_accessor
    celltypes[pinname] = wa.celltype, wa.subcelltype
cachemanager = manager.cachemanager
transformation_cache = cachemanager.transformation_cache
outputname = ctx.tf._output_name
outputpin0 = ctx.tf._pins[outputname]
output_celltype = outputpin0.celltype
output_subcelltype = outputpin0.subcelltype
outputpin = outputname, output_celltype, output_subcelltype

tf_cache = ctx._get_manager().cachemanager.transformation_cache
async def main():
    transformation, transformation_build_exception = await tf_cache.build_transformation(
        ctx.tf, celltypes,inputpin_checksums,outputpin
    )
    meta_checksum = transformation["__meta__"]
    meta = json.loads(buffer_cache.get_buffer(meta_checksum))
    print(meta)

import asyncio
asyncio.get_event_loop().run_until_complete(
    main()
)    