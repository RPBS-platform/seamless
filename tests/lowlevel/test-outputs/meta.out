{'calculation_time': '1s', 'transformer_path': ['tf']}
